### PR TITLE
[runtime env] runtime env inheritance refactor

### PR DESCRIPTION
**Pull request by SongGuyang**
## Why are these changes needed?

Recreate for the reverted PR for 2.0. 
https://github.com/ray-project/ray/pull/22244
https://github.com/ray-project/ray/pull/22620

Runtime Environments is already GA in Ray 1.6.0. The latest doc is [here](https://docs.ray.io/en/master/ray-core/handling-dependencies.html#runtime-environments). And now, we already supported a [inheritance](https://docs.ray.io/en/master/ray-core/handling-dependencies.html#inheritance) behavior as follows (copied from the doc):
- The runtime_env["env_vars"] field will be merged with the runtime_env["env_vars"] field of the parent. This allows for environment variables set in the parent’s runtime environment to be automatically propagated to the child, even if new environment variables are set in the child’s runtime environment.
- Every other field in the runtime_env will be overridden by the child, not merged. For example, if runtime_env["py_modules"] is specified, it will replace the runtime_env["py_modules"] field of the parent.

We think this runtime env merging logic is so complex and confusing to users because users can't know the final runtime env before the jobs are run.

Current PR tries to do a refactor and change the behavior of Runtime Environments inheritance. Here is the new behavior:
- **If there is no runtime env option when we create actor, inherit the parent runtime env.**
- **Otherwise, use the optional runtime env directly and don't do the merging.**

You could get the parent runtime env and modify this dict by yourself. Like:
```Actor.options(runtime_env=ray.get_runtime_context().runtime_env.update({"X": "Y"}))```

## Related issue number

https://github.com/ray-project/ray/issues/21818

### CODE REVIEW

**Comment by SongGuyang** [comment]
Hey @xwjiang2010, we will make this change again.  Do you already have one UT for the issue https://github.com/ray-project/ray/issues/22595?

**Comment by SongGuyang** [comment]
@edoakes @rkooo567 @architkulkarni please help to check if we can make this change for 2.0.

**Comment by rkooo567** [comment]
cc @edoakes Sorry that i forgot some context. did it pass the API review btw? Should we add an api-review-required tag?

**Comment by architkulkarni** [comment]
> cc @edoakes Sorry that i forgot some context. did it pass the API review btw? Should we add an api-review-required tag?

This was the API review issue: https://github.com/ray-project/ray/issues/21818 There wasn't any pushback.  We also asked for opinions on OSS Slack and I think I remember there were a few thumbs up and no pushback (unfortunately I think the thread is lost because we have limited history.)

**Review by architkulkarni — APPROVED** [review]
The two PRs mentioned were already approved, so this PR should be approved too.  It includes the warning message for users in the case where the behavior is different.  The Ray 1.13 branch cut has already happened, so if I understand correctly, it's safe to merge this now and this will only appear in Ray 2.0.

I'll try to run `train_pytorch_linear_test` directly on this PR, since now we can run release tests for PRs on buildkite. (running [here](https://buildkite.com/ray-project/release-tests-pr/builds/2441))

**Review by edoakes — CHANGES_REQUESTED** [review]
According to @zhe-thoughts we should wait for the `branch-1` to be cut before merging breaking changes, so holding for that.

**Review by rkooo567 — APPROVED** [review]
Approving the PR assuming I already approved the same PR before!

**Comment by SongGuyang** [comment]
We will merge this PR because the "ray-1.x" branch has been cut.

File: java/test/src/main/java/io/ray/test/RuntimeEnvTest.java
Comment (rkooo567) [reply java/test/src/main/java/io/ray/test/RuntimeEnvTest.java:58]: Let's merge this PR after addressing this? @jovany-wang we can just change the code this way here?

Comment (jovany-wang) [reply java/test/src/main/java/io/ray/test/RuntimeEnvTest.java:58]: This change had been committed at [867ef68](https://github.com/ray-project/ray/pull/24538/commits/867ef6825597c6b1c62449519727c0c271f6f5de)

So feel free to merge this PR

**Review by jovany-wang — APPROVED** [review]

**Review by edoakes — APPROVED** [review]

**Comment by ericl** [comment]
This PR has caused a P0 regression in master for Tune: https://github.com/ray-project/ray/issues/25484

Thinking about the new behavior, it's also fundamentally inconsistent with how Jobs works. When the user specifies a working_dir, it's very surprising that overriding env_vars also deletes the working dir.

I think we need to change this to either:
1. Raise an *exception* if any fields are dropped (such as working_dir), or
2. Revert to the previous behavior.

The user should have to explicitly set `{"working_dir": None}` to unset the working_dir.

**Pull request merged**
